### ISSUE TITLE
Crashing when trying to download book

### ISSUE THREAD

**Issue by Ninjalord-25** [rmcrackan/Libation#294]
when I go into download a book, it starts up like normal, click the red traffic light to start the process, it shows up in the download bar and then briefly flashes to the time to wait for completion, then the program promptly closes without having done a thing
[Log202206_003.log](https://github.com/rmcrackan/Libation/files/8970134/Log202206_003.log)

**Comment by Foardskin** [comment]
I am having the same problem, crashing as soon as the download starts in the process queue. But that folder you mentioned doesn't exist on my computer

**Comment by Mbucari** [comment]
@Foardskin Please attach you log file

**Comment by Foardskin** [comment]
I don't know how to find one of those sorry. 
the log in libation, on the the tab next to the process queue is blank.

**Comment by Mbucari** [comment]
In Libation top tool menu, click Settings > Settings...

The Setings window will open up.  On that first page you'll see a button called "Open log Folder". Click it. You'll see at least one file in there whose name begins with "Log". Upload all such files here (just drag-drop the files into the github text editor.

**Comment by Mbucari** [comment]
@Foardskin That log file is very old. Are there any other log files that don't begin with an underscore (_)?

**Comment by Foardskin** [comment]
[Log202206.log](https://github.com/rmcrackan/Libation/files/8970674/Log202206.log)

here we go, thank you

**Comment by Foardskin** [comment]
I found the downloadinprogress folder. AppData was a hidden folder in windows 10.

unfortunately I'm getting the same issue with libation closing on download start

[Log202206.log](https://github.com/rmcrackan/Libation/files/8970754/Log202206.log)
t

**Comment by Mbucari** [comment]
@rmcrackan I think the only thing we can do for not is to add more logging and error handling.  There's just not enough info in these reports to identify a problem.

I'm working on that now and will submit a PR tonight for you to review tomorrow.

**Comment by Ninjalord-25** [comment]
> So sorry about the crashing. Thanks for the log. Please delete the files in `C:\Users\chaot\AppData\Local\Temp\Libation\DownloadsInProgress` and try again. If you're still crashing, please attach the new log file.

did that and it happened the same as before, here's the new log file
[Log202206_003.log](https://github.com/rmcrackan/Libation/files/8971714/Log202206_003.log)

**Comment by rmcrackan** [comment]
Thank you for your patience. See if [version 8.1.2](https://github.com/rmcrackan/Libation/releases/latest) solves your issue. If it doesn't, it should at least give us much better logs to figure this out.

**Comment by Ninjalord-25** [comment]
Nada, i picked a few books at random to try and download all at once if that helps any this time.
[Log202206_003.log](https://github.com/rmcrackan/Libation/files/8972315/Log202206_003.log)

**Comment by rmcrackan** [comment]
Well that's a kick in the teeth. Not only is it failing in the same place, it's not even producing better logs. Stay tuned

**Comment by rmcrackan** [comment]
Can you give examples of some failing books? I'm hoping I have some of the same ones. Especially look for free titles or popular titles which are failing

**Comment by Mbucari** [comment]
@Ninjalord-25 How are you starting your downloads?  Are you:

- Clicking the stoplight icon for individual books?
- Clicking Visible Books > Liberate?
- Clicking Liberate > Begin Book and PDF Backup?
- Clicking Liberate > Liberate Visible Books?

**Comment by Ninjalord-25** [comment]
> Can you give examples of some failing books? I'm hoping I have some of the same ones. Especially look for free titles or popular titles which are failing

i can give you the names of each book i tried to download when this all started if you'll give me a moment

these books i had previously downloaded and then gone to unmark and redownload them again to get the cover art to go in the folders of the books i already had: We Are Legion (We Are Bob) Bobiverse books 1 by Dennis E Taylor, Invisible by James Patterson and David Ellis, The Lost Gate Mithermages book 1 by Orson Scott Card, Broken Homes a Rivers of London novel by Ben Aaronovitch, Foxglove Summer also by Ben Aaronovitch, 

then several times i tried using Orlando People by Alexander C. Kane as a kind of control, keeping using the same book to see if things changed during all these testing and fixes, and just now i tried using one of the new books i purchased that was just added to the list of books it was Ruin A Zombie Apocalypse LitRPG by David Petrie

**Comment by Ninjalord-25** [comment]
> @Ninjalord-25 How are you starting your downloads? Are you:
> 
> * Clicking the stoplight icon for individual books?
> * Clicking Visible Books > Liberate?
> * Clicking Liberate > Begin Book and PDF Backup?
> * Clicking Liberate > Liberate Visible Books?

Stoplight Icon. i like to do them one at a time to keep up with them and rename the files to something more manageable before organizing them in a different folder to keep everything straight, so i do them in batches at a time.

**Comment by rmcrackan** [comment]
I just tried repeatedly with Bobiverse book 1, including turning on the 'save cover art' and 'split into chapters' settings -- no crash on my machine. Jeez this is killing me!

**Comment by Mbucari** [comment]
> no crash on my machine. Jeez this is killing me!

My guess is it's related to his DB and how queries are being executed.

**Comment by rmcrackan** [comment]
More bug fixes and logging in [version 8.1.3](https://github.com/rmcrackan/Libation/releases/latest)

To anyone having this issue who's willing to share sensitive information *privately* please reach out to me via [email](mailto:rmcrackan@gmail.com). This is an "above and beyond" situation and I will not ask anyone specifically to do this -- it's volunteer only and could really help crack this case.

**Comment by Mbucari** [comment]
@Ninjalord-25 Have you had a chance to try downloading those problematic books with Libation 8.1.3?

**Comment by Ninjalord-25** [comment]
Not had a chance. Been at work all day and won't be off any time soon. Will update once I try

**Comment by Ninjalord-25** [comment]
[Log202206_003.log](https://github.com/rmcrackan/Libation/files/8983963/Log202206_003.log)
here's the most recent log. still no change, am i doing something wrong with the way i download the books?

**Comment by Mbucari** [comment]
Can you try something for me?  Delete your entire Libation program files folder (not your settings folder).
Download a fresh copy [from here](https://github.com/rmcrackan/Libation/releases/download/v8.1.3/Libation.8.1.3.zip) and extract it to a new, different location. Open than folder and run Libation.exe.

**Comment by Thor-74** [comment]
@Mbucari: I have the same problem. I tried to debug the current version and it seems that AaxFile.ConvertToMultiMp3Async (=> await mp4File.ProcessAudio) in AaxcDownloadMultiConverter.ConvertToMultiMp3 causes the crash.  
Additional info: if I choose m4b as output format everything works fine.

**Comment by Mbucari** [comment]
@Thor-74 that's the ticket! I'm able to reproduce the error. I'll try to get to it later today.
 
@rmcrackan maybe we should start asking people for their settings file in addition to their log file. And never even occurred to me that this could be a problem only in multi-part mp3s.

**Comment by rmcrackan** [comment]
@Mbucari 

We can certainly ask for it. fwiw I thought about an enhancement to have a 1-click zip which could include the log files, settings files, whatever. I don't like the opacity of that though. The user has to unzip the file to see if they're exposing what they intend to.

I knew this was happening only in multipart because that's the only place where that log text appears. 1) I tried multi-part and still no repro for me. 2) I still can't wrap my brain around why that would crash the app silently. There's a try/catch around that code and also the first step of the global exception handlers is to log an error. Please please please -- how the hell does this crash the app? It's killing me.

**Comment by Mbucari** [comment]
It's an access violation, so probably something native. I was digging into it a little in the wee hours of the morning and making zero progress, so I just went to sleep. I'll take another look now.

**Comment by Mbucari** [comment]
Well, I found it. God I feel stupid.

In AaxDecrypter.AaxcDownloadMultiConverter line # 158

Change:
```C#
=> Callback(currentChapter, splitChapters, newSplitCallback);
```
to
```C#
=> Callback(currentChapter, splitChapters, newSplitCallback as NewSplitCallback);
```

I have no idea why this wasn't being thrown as a stack overflow error, but there you go.

**Comment by rmcrackan** [comment]
Good find. As per email: I put your fix in place. Email me when you're ready for me to publish. fyi my availability will be hit and miss this weekend.

**Comment by rmcrackan** [comment]
Fixed in [version 8.1.4](https://github.com/rmcrackan/Libation/releases/latest). Thanks everyone for working with us on this!

**Comment by Ninjalord-25** [comment]
huzzah! i can download books again. thank yall!
